### PR TITLE
fix: local `cloudProvider` config is overwritten by a central config

### PR DESCRIPTION
**Pull request by trentm**
Fixes: #1976

### Checklist

- [x] Implement code
- [x] Add tests
- [x] Add CHANGELOG.asciidoc entry
- [x] Commit message follows [commit guidelines](https://github.com/elastic/apm-agent-nodejs/blob/master/CONTRIBUTING.md#commit-message-guidelines)

### CODE REVIEW

**Comment by trentm** [comment]
This shows the test failing without the lib/config.js change:

```
% ./node_modules/.bin/tape test/central-config-enabled.js
TAP version 13
...
# central config change does not erroneously update cloudProvider
ok 20 immediately after .start(), cloudProvider=aws
ok 21 should be strictly equal
not ok 22 after fetching central config, cloudProvider is not reset to default
  ---
    operator: equal
    expected: 'aws'
    actual:   'auto'
    at: Client.<anonymous> (/Users/trentm/el/apm-agent-nodejs5/test/central-config-enabled.js:183:9)
    stack: |-
      Error: after fetching central config, cloudProvider is not reset to default
          at Test.assert [as _assert] (/Users/trentm/el/apm-agent-nodejs5/node_modules/tape/lib/test.js:260:54)
          at Test.bound [as _assert] (/Users/trentm/el/apm-agent-nodejs5/node_modules/tape/lib/test.js:84:32)
          at Test.strictEqual (/Users/trentm/el/apm-agent-nodejs5/node_modules/tape/lib/test.js:424:10)
          at Test.bound [as equal] (/Users/trentm/el/apm-agent-nodejs5/node_modules/tape/lib/test.js:84:32)
          at Client.<anonymous> (/Users/trentm/el/apm-agent-nodejs5/test/central-config-enabled.js:183:9)
          at Object.onceWrapper (events.js:286:20)
          at Client.emit (events.js:203:15)
          at streamToBuffer (/Users/trentm/el/apm-agent-nodejs5/node_modules/elastic-apm-http-client/index.js:191:16)
          at IncomingMessage.<anonymous> (/Users/trentm/el/apm-agent-nodejs5/node_modules/fast-stream-to-buffer/index.js:20:9)
          at IncomingMessage.f (/Users/trentm/el/apm-agent-nodejs5/node_modules/once/once.js:25:25)
          at IncomingMessage.onend (/Users/trentm/el/apm-agent-nodejs5/node_modules/end-of-stream/index.js:36:27)
          at IncomingMessage.emit (events.js:203:15)
          at endReadableNT (_stream_readable.js:1145:12)
          at process._tickCallback (internal/process/next_tick.js:63:19)
  ...

1..22
# tests 22
# pass  21
# fail  1
```


and passes afterwards:

```
# central config change does not erroneously update cloudProvider
ok 20 immediately after .start(), cloudProvider=aws
ok 21 should be strictly equal
ok 22 after fetching central config, cloudProvider is not reset to default

1..22
# tests 22
# pass  22

# ok
```


This also updates the logger.warning to correctly log the invalid given value.

```
% ELASTIC_APM_LOG_LEVEL=trace node esapp.js
Invalid "cloudProvider" config value boguscloud, falling back to default auto
agent configured correctly { pid: 58528,
  ppid: 58382,
  arch: 'x64',
...
     cloudProvider: 'auto',
...
```

I've also started explicitly passing the logger through to the internal functions that might log, to avoid having to guard against `opts.logger` not existing.

File: lib/config.js [hunk 1]
```diff
@@ -435,15 +435,13 @@ function normalizeSanitizeFieldNames (opts) {
...
-        'Unexpected cloudProvider value %s, changing to auto',
-        opts.cloudProvider
-      )
+function normalizeCloudProvider (opts, logger) {
+  if ('cloudProvider' in opts) {
+    const allowedValues = ['auto', 'gcp', 'azure', 'aws', 'none']
+    if (allowedValues.indexOf(opts.cloudProvider) === -1) {
+      logger.warn('Invalid "cloudProvider" config value %s, falling back to default %s',
+        opts.cloudProvider, DEFAULTS.cloudProvider)
+      opts.cloudProvider = DEFAULTS.cloudProvider
```
Comment (astorm) [inline lib/config.js:444]: I see here we've changed from using a hard coded `auto` to using the value set in `DEFAULTS` -- 👍 future proofing.

File: lib/config.js [hunk 2]
```diff
@@ -435,15 +435,13 @@ function normalizeSanitizeFieldNames (opts) {
...
-  if (allowedValues.indexOf(opts.cloudProvider) === -1) {
-    opts.cloudProvider = 'auto'
-    if (opts.logger) {
-      opts.logger.warn(
-        'Unexpected cloudProvider value %s, changing to auto',
-        opts.cloudProvider
-      )
+function normalizeCloudProvider (opts, logger) {
+  if ('cloudProvider' in opts) {
+    const allowedValues = ['auto', 'gcp', 'azure', 'aws', 'none']
```
Comment (astorm) [inline lib/config.js:440]: I see here we're only normalizing the value if there's an actual property on the `opts` array. My presumption is that it's this line that prevents the the central configuration update from triggering an update of the `cloudProvider` value.

**Review by astorm — APPROVED** [review]
👍 Does the thing and the secondary changes w/r/t the additional logger argument are good as well.  Approving.

**Pull request merged**